In JUnit XML reports, the timestamp attribute is not populated correctly for all subTests

Hello,

I have been experimenting with nose2 JUnit XML output plugin in conjunction with `unittest` `subTest` context manager and found that the `timestamp` attribute is populated correctly only for the first generated `testcase` node. For the subsequent nodes, the timestamp is set to '1970-01-01T00:00:00'.

A trivial example:
```python

class SubTestReportTest(unittest.TestCase):
    def test_subtests(self):
        for cnt in range(10):
            with self.subTest("Testing cnt = {}".format(cnt), cnt=cnt):
                self.assertLess(cnt, 4, "Verify that the cnt is < 4")
```

... yields the following XML output:
```xml
<testsuite errors="0" failures="6" name="nose2-junit" skipped="0" tests="1" time="0.001">
  <testcase classname="__main__.SubTestReportTest" name="test_subtests [Testing cnt = 4] (cnt=4)" time="0.000497" timestamp="2023-04-05T09:54:42.927951">
    <failure message="test failure">Traceback (most recent call last):
  File "./test_reports.py", line 47, in test_subtests
    self.assertLess(cnt, 4, "Verify that the cnt is &lt; 4")
AssertionError: 4 not less than 4 : Verify that the cnt is &lt; 4
</failure>
    <system-out />
  </testcase>
  <testcase classname="__main__.SubTestReportTest" name="test_subtests [Testing cnt = 5] (cnt=5)" time="0.000000" timestamp="1970-01-01T00:00:00">
    <failure message="test failure">Traceback (most recent call last):
  File "./test_reports.py", line 47, in test_subtests
    self.assertLess(cnt, 4, "Verify that the cnt is &lt; 4")
AssertionError: 5 not less than 4 : Verify that the cnt is &lt; 4
</failure>
    <system-out />
  </testcase>
  <testcase classname="__main__.SubTestReportTest" name="test_subtests [Testing cnt = 6] (cnt=6)" time="0.000000" timestamp="1970-01-01T00:00:00">
    <failure message="test failure">Traceback (most recent call last):
  File "./test_reports.py", line 47, in test_subtests
    self.assertLess(cnt, 4, "Verify that the cnt is &lt; 4")
AssertionError: 6 not less than 4 : Verify that the cnt is &lt; 4
</failure>
    <system-out />
  </testcase>
  <testcase classname="__main__.SubTestReportTest" name="test_subtests [Testing cnt = 7] (cnt=7)" time="0.000000" timestamp="1970-01-01T00:00:00">
    <failure message="test failure">Traceback (most recent call last):
  File "./test_reports.py", line 47, in test_subtests
    self.assertLess(cnt, 4, "Verify that the cnt is &lt; 4")
AssertionError: 7 not less than 4 : Verify that the cnt is &lt; 4
</failure>
    <system-out />
  </testcase>
  <testcase classname="__main__.SubTestReportTest" name="test_subtests [Testing cnt = 8] (cnt=8)" time="0.000000" timestamp="1970-01-01T00:00:00">
    <failure message="test failure">Traceback (most recent call last):
  File "./test_reports.py", line 47, in test_subtests
    self.assertLess(cnt, 4, "Verify that the cnt is &lt; 4")
AssertionError: 8 not less than 4 : Verify that the cnt is &lt; 4
</failure>
    <system-out />
  </testcase>
  <testcase classname="__main__.SubTestReportTest" name="test_subtests [Testing cnt = 9] (cnt=9)" time="0.000000" timestamp="1970-01-01T00:00:00">
    <failure message="test failure">Traceback (most recent call last):
  File "./test_reports.py", line 47, in test_subtests
    self.assertLess(cnt, 4, "Verify that the cnt is &lt; 4")
AssertionError: 9 not less than 4 : Verify that the cnt is &lt; 4
</failure>
    <system-out />
  </testcase>
</testsuite>
```
Ideally, the expected output would be for each failure report to be associated with a timestamp indicating when the failure occurred. Alternatively, it could be a timestamp taken when the test case containing the `subTest`s was started.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -13,6 +13,12 @@ testsuites.
 Unreleased
 ----------
 
+Fixed
+~~~~~
+
+* JUnit XML reports now use the containing test's start timestamp for every
+  failing subtest. (:issue:`567`)
+
 0.16.0 (2026-03-01)
 -------------------
 

--- a/nose2/plugins/junitxml.py
+++ b/nose2/plugins/junitxml.py
@@ -1,9 +1,9 @@
 """
 Output test reports in junit-xml format.
 
-This plugin implements :func:`startTest`, :func:`testOutcome` and
-:func:`stopTestRun` to compile and then output a test report in
-junit-xml format. By default, the report is written to a file called
+This plugin implements :func:`startTest`, :func:`stopTest`,
+:func:`testOutcome` and :func:`stopTestRun` to compile and then output a test
+report in junit-xml format. By default, the report is written to a file called
 ``nose2-junit.xml`` in the current working directory.
 
 You can configure the output filename by setting ``path`` in a ``[junit-xml]``
@@ -145,6 +145,10 @@ class JUnitXmlReporter(events.Plugin):
         """Count test, record start time"""
         self.numtests += 1
         self._start = event.startTime
+
+    def stopTest(self, event):
+        """Clear the start time after all outcomes for a test are reported."""
+        self._start = None
 
     def testOutcome(self, event):
         """Add test outcome to xml tree"""
@@ -293,10 +297,7 @@ class JUnitXmlReporter(events.Plugin):
         try:
             return time.time() - self._start
         except Exception:
-            pass
-        finally:
-            self._start = None
-        return 0
+            return 0
 
     def _iso_timestamp(self):
         # in some cases, `self._start` is still None when this runs, convert to 0

--- a/nose2/tests/unit/test_junitxml.py
+++ b/nose2/tests/unit/test_junitxml.py
@@ -201,6 +201,22 @@ class TestJunitXmlPlugin(TestCase):
         self.assertIsNone(skip.get("message"))
         self.assertIsNone(skip.text)
 
+    def test_subtest_timestamps_use_parent_start(self):
+        class SubTestCase(unittest.TestCase):
+            def test_failing_subtests(self):
+                for value in range(2):
+                    with self.subTest(value=value):
+                        self.fail()
+
+        test = SubTestCase("test_failing_subtests")
+        test(self.result)
+        timestamps = [
+            case.get("timestamp") for case in self.plugin.tree.findall("testcase")
+        ]
+        self.assertEqual(len(timestamps), 2)
+        self.assertEqual(timestamps[0], timestamps[1])
+        self.assertNotEqual(timestamps[0], "1970-01-01T00:00:00")
+
     def test_generator_timestamp_increases(self):
         gen = generators.Generators(session=self.session)
         gen.register()

